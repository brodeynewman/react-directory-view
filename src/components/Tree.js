import _ from 'lodash';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import TreeNode from './TreeNode';
import DefaultComponent from './defaultComponent/Default';

/**
 * Recursively maps over a tree, and renders tree nodes
 * @param {Object} props - component props
 * @param {Object} tree - current tree node
 */
const mapChildrenRecursively = (props, tree, paddingLeft = 0) => {
  if (tree.length) {
    return _.map(tree, child => (
      <TreeNode
        key={_.get(child, `${_.get(props, 'treeMap.nodeKey', '')}`)}
        child={_.get(child, `${_.get(props, 'treeMap.childToRender', '')}`)}
        paddingLeft={paddingLeft}
        isDeepestChild={!_.get(child, `${_.get(props, 'treeMap.recursiveKey', 'children')}.length`)}
        onExpand={_.get(props, 'treeMap.onExpand', _.noop)}
        onContract={_.get(props, 'treeMap.onContract', _.noop)}
        ComponentToRender={_.get(props, 'treeMap.Component', DefaultComponent)}
        useCheckbox={_.get(props, 'treeMap.useCheckbox', false)}
        checkboxStyles={_.get(props, 'checkboxStyles', {})}
        arrowStyles={_.get(props, 'arrowStyles', {})}
        {...child}
      >
        {
            mapChildrenRecursively(
              props,
              _.get(child, `${_.get(props, 'treeMap.recursiveKey', 'children')}`, []),
              paddingLeft + (_.get(props, 'treeMap.paddingLeft', 15)),
            )
          }
      </TreeNode>
    ));
  }

  return null;
};

/**
 * Factory function used for partialing
 * @param {Object} props - component props
 */
const treeFactory = props => ({
  mapChildrenRecursively: _.partial(mapChildrenRecursively, props),
});

class Tree extends Component {
  constructor(props) {
    super(props);

    this.handleClick = this.handleClick.bind(this);
  }

  handleClick() {
    this.setState(prevState => ({ collapseAll: !prevState.collapseAll }));
  }

  render() {
    const { treeData } = this.props;

    const factoryWithProps = treeFactory(this.props);

    return (
      <div>
        {
          treeData ?
          factoryWithProps.mapChildrenRecursively(treeData)
          : null
        }
      </div>
    );
  }
}

Tree.propTypes = {
  treeData: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default Tree;
