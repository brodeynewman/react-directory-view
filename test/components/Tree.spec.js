import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Tree from '../../src/components/Tree';

const treeData = [
  {
    path: 'this',
    id: 1,
    children: [
      {
        path: 'is',
        id: 2,
        children: [
          {
            path: 'a',
            id: 3,
            children: [
              {
                path: 'few',
                id: 4,
                children: [
                  {
                    path: 'deep',
                    id: 5,
                    children: [],
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
];

describe('Tree', () => {
  it('Tree renders correctly', () => {
    const tree = renderer
      .create(<Tree treeData={treeData} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Tree renders a div if treeData is non-existent', () => {
    const tree = renderer
      .create(<Tree />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
