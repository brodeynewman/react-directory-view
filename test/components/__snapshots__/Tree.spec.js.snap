// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Tree renders a div if treeData is non-existent 1`] = `<div />`;

exports[`Tree Tree renders correctly 1`] = `
<div>
  <div>
    <div>
      <span
        style={
          Object {
            "paddingLeft": 0,
          }
        }
      >
        <span
          onClick={[Function]}
          role="presentation"
          style={
            Object {
              "border": "4px solid transparent",
              "borderRight": "4px solid",
              "borderTop": "4px solid",
              "boxSizing": "border-box",
              "color": "#666",
              "cursor": "pointer",
              "display": "inline-block",
              "height": 0,
              "margin": 15,
              "position": "relative",
              "transform": "rotate(45deg)",
              "verticalAlign": "middle",
              "width": 0,
            }
          }
        />
      </span>
      
    </div>
  </div>
  <div
    style={
      Object {
        "display": "none",
        "visibility": "hidden",
      }
    }
  >
    <div>
      <div>
        <span
          style={
            Object {
              "paddingLeft": 15,
            }
          }
        >
          <span
            onClick={[Function]}
            role="presentation"
            style={
              Object {
                "border": "4px solid transparent",
                "borderRight": "4px solid",
                "borderTop": "4px solid",
                "boxSizing": "border-box",
                "color": "#666",
                "cursor": "pointer",
                "display": "inline-block",
                "height": 0,
                "margin": 15,
                "position": "relative",
                "transform": "rotate(45deg)",
                "verticalAlign": "middle",
                "width": 0,
              }
            }
          />
        </span>
        
      </div>
    </div>
    <div
      style={
        Object {
          "display": "none",
          "visibility": "hidden",
        }
      }
    >
      <div>
        <div>
          <span
            style={
              Object {
                "paddingLeft": 30,
              }
            }
          >
            <span
              onClick={[Function]}
              role="presentation"
              style={
                Object {
                  "border": "4px solid transparent",
                  "borderRight": "4px solid",
                  "borderTop": "4px solid",
                  "boxSizing": "border-box",
                  "color": "#666",
                  "cursor": "pointer",
                  "display": "inline-block",
                  "height": 0,
                  "margin": 15,
                  "position": "relative",
                  "transform": "rotate(45deg)",
                  "verticalAlign": "middle",
                  "width": 0,
                }
              }
            />
          </span>
          
        </div>
      </div>
      <div
        style={
          Object {
            "display": "none",
            "visibility": "hidden",
          }
        }
      >
        <div>
          <div>
            <span
              style={
                Object {
                  "paddingLeft": 45,
                }
              }
            >
              <span
                onClick={[Function]}
                role="presentation"
                style={
                  Object {
                    "border": "4px solid transparent",
                    "borderRight": "4px solid",
                    "borderTop": "4px solid",
                    "boxSizing": "border-box",
                    "color": "#666",
                    "cursor": "pointer",
                    "display": "inline-block",
                    "height": 0,
                    "margin": 15,
                    "position": "relative",
                    "transform": "rotate(45deg)",
                    "verticalAlign": "middle",
                    "width": 0,
                  }
                }
              />
            </span>
            
          </div>
        </div>
        <div
          style={
            Object {
              "display": "none",
              "visibility": "hidden",
            }
          }
        >
          <div>
            <div>
              <span
                style={
                  Object {
                    "paddingLeft": 60,
                  }
                }
              />
              
            </div>
          </div>
          <div
            style={
              Object {
                "display": "none",
                "visibility": "hidden",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
